8212879: Make JVMTI TagMap table concurrent

coleenp opened the pull request:
This change turns the HashTable that JVMTI uses for object tagging into a regular Hotspot hashtable - the one in hashtable.hpp with resizing and rehashing.   Instead of pointing directly to oops so that GC has to walk the table to follow oops and then to rehash the table, this table points to WeakHandle.  GC walks the backing OopStorages concurrently.

The hash function for the table is a hash of the lower 32 bits of the address.  A flag is set during GC (gc_notification if in a safepoint, and through a call to JvmtiTagMap::needs_processing()) so that the table is rehashed at the next use.

The gc_notification mechanism of weak oop processing is used to notify Jvmti to post ObjectFree events.  In concurrent GCs there can be a window of time between weak oop marking where the oop is unmarked, so dead (the phantom load in peek returns NULL) but the gc_notification hasn't been done yet.  In this window, a heap walk or GetObjectsWithTags call would not find an object before the ObjectFree event is posted.  This is dealt with in two ways:

1. In the Heap walk, there's an unconditional table walk to post events if events are needed to post.
2. For GetObjectWithTags, if a dead oop is found in the table and posting is required, we use the VM thread to post the event.

Event posting cannot be done in a JavaThread because the posting needs to be done while holding the table lock, so that the JvmtiEnv state doesn't change before posting is done.  ObjectFree callbacks are limited in what they can do as per the JVMTI Specification.  The allowed callbacks to the VM already have code to allow NonJava threads.

To avoid rehashing, I also tried to use object->identity_hash() but this breaks because entries can be added to the table during heapwalk, where the objects use marking.  The starting markWord is saved and restored.  Adding a hashcode during this operation makes restoring the former markWord (locked, inflated, etc) too complicated.  Plus we don't want all these objects to have hashcodes because locking operations after tagging would have to always use inflated locks.

Much of this change is to remove serial weak oop processing for the weakProcessor, ZGC and Shenandoah.  The GCs have been stress tested with jvmti code.

It has also been tested with tier1-6.

Thank you to Stefan, Erik and Kim for their help with this change.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux aarch64 | Linux arm | Linux ppc64le | Linux s390x | Linux x64 | Linux x86 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- | ----- | ----- | ----- | ----- | ----- |
| Build | ⏳ (1/1 running) | ⏳ (1/1 running) | ⏳ (1/1 running) | ⏳ (1/1 running) | ✔️ (6/6 passed) | ✔️ (2/2 passed) | ⏳ (1/2 running) | ✔️ (2/2 passed) |
| Test (tier1) |    |     |     |     |  ⏳ (9/9 running) | ⏳ (4/9 running) |    |  ⏳ (9/9 running) |

### Issue
 * [JDK-8212879](https://bugs.openjdk.java.net/browse/JDK-8212879): Make JVMTI TagMap table concurrent


### Reviewers
 * [Stefan Karlsson](https://openjdk.java.net/census#stefank) (@stefank - **Reviewer**) ⚠️ Review applies to cb4c83e0cd89f737be647e3bc8f336f471e2eebc
 * [Magnus Ihse Bursie](https://openjdk.java.net/census#ihse) (@magicus - **Reviewer**) ⚠️ Review applies to 534326dacce929013b3c7063258f2bf4f31b1024
 * [Zhengyu Gu](https://openjdk.java.net/census#zgu) (@zhengyu123 - **Reviewer**) ⚠️ Review applies to cb4c83e0cd89f737be647e3bc8f336f471e2eebc
 * [Erik Österlund](https://openjdk.java.net/census#eosterlund) (@fisk - **Reviewer**) ⚠️ Review applies to eeaf9aed5a696b0e1b6b2cfcd8f6ee93f0210710
 * [Serguei Spitsyn](https://openjdk.java.net/census#sspitsyn) (@sspitsyn - **Reviewer**) ⚠️ Review applies to 1940eaf16287b2618e34a7f47e5adb21f2afa1be
 * [Kim Barrett](https://openjdk.java.net/census#kbarrett) (@kimbarrett - **Reviewer**) ⚠️ Review applies to 1940eaf16287b2618e34a7f47e5adb21f2afa1be


### Contributors
 * Kim Barrett `<kbarrett@openjdk.org>`
 * Coleen Phillimore `<coleenp@openjdk.org>`

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/967/head:pull/967`
`$ git checkout pull/967`

erikj79 reviewed (commented):
Build changes look ok.

stefank commented on src/hotspot/share/prims/jvmtiTagMapTable.cpp:
```diff
@@ -0,0 +1,269 @@
+  unsigned int hash = compute_hash(obj);
+  int index = hash_to_index(hash);
+  // One was added while acquiring the lock
+  JvmtiTagMapEntry* entry = find(index, hash, obj);
```
Should this be done inside ASSERT?

coleenp replied:
yes.

stefank commented on test/hotspot/jtreg/vmTestbase/nsk/jvmti/scenarios/capability/CM02/cm02t001/cm02t001.cpp:
```diff
@@ -60,6 +60,10 @@ static jlong timeout = 0;
 static jthread thread = NULL;
 static jclass klass = NULL;
+static jobject testedObject = NULL;
+const jlong TESTED_TAG_VALUE = (5555555L);
```
Remove parenthesis?

coleenp replied:
I copied this from some other place that had parenthesis.

stefank reviewed (commented):
Commented on nits, and reviewed GC code and tag map code. Didn't look closely on the hashmap changes.

magicus reviewed (approved):
Build changes look good. Not reviewed hotspot changes.

coleenp commented on src/hotspot/share/prims/jvmtiTagMap.cpp:
```diff
@@ -3298,117 +2972,57 @@ void JvmtiTagMap::follow_references(jint heap_filter,
-        tag_map->do_weak_oops(is_alive, f);
-      }
+  // Hopefully quick exit.
+  if (!JvmtiEnv::environments_might_exist()) { return; }
+
+  JvmtiEnvIterator it;
+  for (JvmtiEnv* env = it.first(); env != NULL; env = it.next(env)) {
```
yes, it looks like it does the same thing.  I'll remove it.

coleenp commented on src/hotspot/share/prims/jvmtiTagMap.cpp:
```diff
@@ -3298,117 +2972,57 @@ void JvmtiTagMap::follow_references(jint heap_filter,
-        // post the event to the profiler
-        if (post_object_free) {
-          JvmtiExport::post_object_free(env(), tag);
+  // No locks during VM bring-up (0 threads) and no safepoints after main
+  // thread creation and before VMThread creation (1 thread); initial GC
+  // verification can happen in that window which gets to here.
+  if (!JvmtiEnv::environments_might_exist()) { return; }
```
I've spent tons of time trying to understand this comment too. I think gc verification used to call oops do on the tagmap table.  This comments obsolete now, and I'll remove it.

coleenp commented on src/hotspot/share/prims/jvmtiTagMap.cpp:
```diff
@@ -3298,117 +2972,57 @@ void JvmtiTagMap::follow_references(jint heap_filter,
+    for (JvmtiEnv* env = it.first(); env != NULL; env = it.next(env)) {
+      JvmtiTagMap* tag_map = env->tag_map_acquire();
+      if (tag_map != NULL && !tag_map->is_empty()) {
+        if (num_dead_entries > 0) {
```
ok.

coleenp commented on src/hotspot/share/prims/jvmtiTagMap.cpp:
```diff
@@ -3298,117 +2972,57 @@ void JvmtiTagMap::follow_references(jint heap_filter,
+        if (num_dead_entries > 0) {
+          tag_map->hashmap()->unlink_and_post(tag_map->env());
         }
+        tag_map->_needs_rehashing = true;
```
// Later GC code will relocate the oops, so defer rehashing until then.
?

coleenp commented on src/hotspot/share/prims/jvmtiTagMap.hpp:
```diff
@@ -58,21 +49,16 @@ class JvmtiTagMap :  public CHeapObj<mtInternal> {
 
+  void entry_iterate(JvmtiTagMapEntryClosure* closure);
+  void post_dead_object_on_vm_thread();
  public:
```
ok

coleenp commented on src/hotspot/share/prims/jvmtiTagMap.hpp:
```diff
@@ -120,8 +106,12 @@ class JvmtiTagMap :  public CHeapObj<mtInternal> {
+
+  void unlink_and_post_locked();
+
+  static void check_hashmaps_for_heapwalk();
+  static void set_needs_processing() NOT_JVMTI_RETURN;
+  static void gc_notification(size_t num_dead_entries) NOT_JVMTI_RETURN;
```
I will do (might have already done) that.  Building non-oracle platforms builds minimal vm.

coleenp commented on src/hotspot/share/prims/jvmtiTagMapTable.cpp:
```diff
@@ -0,0 +1,269 @@
+}
+
+oop JvmtiTagMapEntry::object_no_keepalive() {
+  // The AS_NO_KEEPALIVE peeks at the oop without keeping it alive.
+  // This is dangerous in general but is okay if the loaded oop does
+  // not leak out past a thread transition where a safepoint can happen.
+  // A subsequent oop_load without AS_NO_KEEPALIVE (the object() accessor)
+  // keeps the oop alive before doing so.
+  return literal().peek();
```
I'm going to say:
// Just peek at the object without keeping it alive.

coleenp commented on src/hotspot/share/prims/jvmtiTagMapTable.cpp:
```diff
@@ -0,0 +1,269 @@
+
+void JvmtiTagMapTable::free_entry(JvmtiTagMapEntry* entry) {
+  unlink_entry(entry);
+  entry->literal().release(JvmtiExport::weak_tag_storage()); // release OopStorage
```
fixed

coleenp commented on src/hotspot/share/prims/jvmtiTagMapTable.cpp:
```diff
@@ -0,0 +1,269 @@
+      oop target = p->object_no_keepalive();
+
+      // The obj is in the table as a target already
+      if (target != NULL && target == obj) {
```
makes sense.

coleenp reviewed (commented):
I think I addressed your comments, retesting now.  Thank you!

coleenp commented:
There should be a thank you emoji that doesn't send email except maybe to the person reviewing the code. Thank you @erikj79 and @magicus for reviewing the build changes.  There should also be a 'fixed' emoji.

zhengyu123 reviewed (approved):
Shenandoah part looks good.

fisk commented on src/hotspot/share/prims/jvmtiTagMap.cpp:
```diff
@@ -3298,117 +2972,57 @@ void JvmtiTagMap::follow_references(jint heap_filter,
+        if (tag_map != NULL) {
+          // Lock each hashmap from concurrent posting and cleaning
+          MutexLocker ml(tag_map->lock(), Mutex::_no_safepoint_check_flag);
+          tag_map->hashmap()->unlink_and_post(tag_map->env());
```
This could call unlink_and_post_locked instead of manually locking.

coleenp replied:
Ok, 2 requests.  I can call that then and move the logging.

fisk reviewed (changes requested):
Looks great in general. Great work Coleen, and thanks again for fixing this. I like all the red lines in the GC code. I added a few nits/questions.

coleenp commented on test/hotspot/jtreg/vmTestbase/nsk/jvmti/scenarios/capability/CM02/cm02t001/cm02t001.cpp:
```diff
@@ -580,6 +653,7 @@ static int checkGeneratedEvents() {
     if (!NSK_VERIFY(GarbageCollectionFinishEventsCount != 0))
         result = NSK_FALSE;
 
+    printf("Object free events %d\n", ObjectFreeEventsCount);
```
yes, removed it.

coleenp commented on src/hotspot/share/prims/jvmtiTagMap.cpp:
```diff
@@ -731,20 +339,22 @@ class TwoOopCallbackWrapper : public CallbackWrapper {
+  MutexLocker ml(lock(), Mutex::_no_safepoint_check_flag);
+
+  // Check if we have to process for concurrent GCs.
+  check_hashmap(false);
```
Ok, will I run afoul of the ZGC people putting the parameter name after the parameter and the rest of the code, it is before?

fisk replied:
ZGC people passionately place the comment after the argument value.

coleenp replied:
I see that but in the non-zgc code it's the opposite and this is non-zgc code.

coleenp commented on src/hotspot/share/prims/jvmtiTagMap.cpp:
```diff
@@ -1331,6 +945,14 @@ void IterateOverHeapObjectClosure::do_object(oop o) {
     return;
   }
+
+  // skip if object is a dormant shared object whose mirror hasn't been loaded
+  if (o != NULL && o->klass()->java_mirror() == NULL) {
+    log_debug(cds, heap)("skipped dormant archived object " INTPTR_FORMAT " (%s)", p2i(o),
+                         o->klass()->external_name());
+    return;
+  }
```
Now I remember.  I added an assert in JvmtiTagMapTable::find() for oop != NULL which didn't exist in the current hashmap code.  The current hashmap code just didn't find a null oop.  I tracked it down to the fact that we're finding dormant objects whose class hasn't been loaded yet.

coleenp replied:
So I think we do load the archived heap from CDS. The heap walker capabilities can be added dynamically.

stefank commented on src/hotspot/share/prims/jvmtiTagMap.cpp:
```diff
@@ -3298,117 +2972,57 @@ void JvmtiTagMap::follow_references(jint heap_filter,
 
+// Concurrent GC needs to call this in relocation pause, so after the oops are moved
+// and have their new addresses, the table can be rehashed.
+void JvmtiTagMap::set_needs_processing() {
```
>  GC callers shouldn't really have to know what processing we're doing here.

I completely disagree with this. It's extremely important that the GC and Runtime code agrees what this code does and where the GC *must* call it. Knowing the details allows us to skip calling this after mark end, but forces us to call it in relocate starte, when objects should start to move. Though, I don't want to block this review because of this point, so if you still thinks that a non-descriptive name is better then we can argue that separately.

coleenp replied:
Ok, I'll rename it to needs_hashing.

stefank commented on src/hotspot/share/prims/jvmtiTagMapTable.cpp:
```diff
@@ -0,0 +1,266 @@
+  return literal().peek();
+}
+
+JvmtiTagMapTable::JvmtiTagMapTable()
+  : Hashtable<WeakHandle, mtServiceability>(_table_size, sizeof(JvmtiTagMapEntry)) {}
+
+
```
Double whitespace

stefank commented on src/hotspot/share/prims/jvmtiTagMapTable.cpp:
```diff
@@ -0,0 +1,266 @@
+}
+
+// Serially remove unused oops from the table, and notify jvmti.
+void JvmtiTagMapTable::unlink_and_post(JvmtiEnv* env) {
+
```
Stray newline

coleenp replied:
that's a useful newline.

stefank commented on src/hotspot/share/prims/jvmtiTagMapTable.cpp:
```diff
@@ -0,0 +1,266 @@
+}
+
+// Rehash oops in the table
+void JvmtiTagMapTable::rehash() {
+
```
Stray newline

stefank commented on src/hotspot/share/prims/jvmtiTagMapTable.hpp:
```diff
@@ -0,0 +1,101 @@
+
+  JvmtiTagMapEntry* find(int index, unsigned int hash, oop obj);
+
+  void resize_if_needed();
+public:
```
Newline between

stefank commented on src/hotspot/share/prims/jvmtiTagMapTable.hpp:
```diff
@@ -0,0 +1,101 @@
+ public:
+  virtual void do_entry(JvmtiTagMapEntry* entry) = 0;
+};
+
+
```
Double newline

stefank commented on src/hotspot/share/prims/jvmtiTagMapTable.cpp:
```diff
@@ -0,0 +1,266 @@
+    }
+  }
+
+  int rehash_len = moved_entries.length();
+  // Now add back in the entries that were removed.
+  for (int i = 0; i < moved_entries.length(); i++) {
```
rehash_len is read, but not used in for loop condition.

coleenp replied:
It's in the logging message below.  I'll use it in the loop too.

stefank commented on src/hotspot/share/prims/jvmtiTagMapTable.cpp:
```diff
@@ -0,0 +1,266 @@
+      closure->do_entry(p);
+    }
+  }
+}
+const int _resize_load_trigger = 5;       // load factor that will trigger the resize
```
Newline between

stefank reviewed (approved):
Some more nit-picking to make the code more consistent.

albertnetymk commented on src/hotspot/share/utilities/hashtable.cpp:
```diff
@@ -136,8 +137,34 @@ static int literal_size(WeakHandle v) {
+  int requested = (int)(_resize_factor*number_of_entries());
+  const int* primelist = use_large_table_sizes ? _large_table_sizes : _small_table_sizes;
+  int arraysize =  use_large_table_sizes ? _large_array_size  : _small_array_size;
+  int newsize = primelist[0];
+  int index = 0;
+  for (newsize = primelist[index]; index < (arraysize - 1);
+       newsize = primelist[++index]) {
+    if (requested <= newsize) {
+      break;
+    }
+  }
+  return newsize;
```
It is existing code, but could be made clearer as part of this PR:
```
  int newsize;
  for (int i=0; i<arraysize; ++i) {
    newsize = primelist[i];
    if (newsize >= requested)
      break;
  }
  return newsize;
```
Additionally, this method could be made `const`, right?

PS: not a review, just a comment in passing

coleenp replied:
Yes, that is a lot simpler and better.  I'd copied that code from another file without changing it that much.

coleenp commented on src/hotspot/share/prims/jvmtiTagMap.cpp:
```diff
@@ -3298,117 +2975,49 @@ void JvmtiTagMap::follow_references(jint heap_filter,
+          tag_map->hashmap()->unlink_and_post(tag_map->env());
         }
+        // Later GC code will relocate the oops, so defer rehashing until then.
+        tag_map->_needs_rehashing = true;
```
Ok, so I'm going to need help to know where in all the different GCs to make this call.

fisk replied:
For what GC is this wrong? I can see that it might yield more work than required, when performing a full GC, but not that it would do too little work. In other words, I can't see how it is wrong, as opposed to inaccurate. Littering GCs with JVMTI hooks so that we can optimize away an operation we do every young GC, from a full GC, does not really seem worth it IMO.

coleenp commented on src/hotspot/share/prims/jvmtiTagMap.cpp:
```diff
@@ -3298,117 +2975,49 @@ void JvmtiTagMap::follow_references(jint heap_filter,
+      JvmtiTagMap* tag_map = env->tag_map_acquire();
+      if (tag_map != NULL && !tag_map->is_empty()) {
+        if (num_dead_entries != 0) {
+          tag_map->hashmap()->unlink_and_post(tag_map->env());
```
The JVMTI code expects the posting to be done quite eagerly presumably during GC, before it has a chance to disable the event or some other such operation.  So the posting is done during the notification because it's as soon as possible.  Deferring to the ServiceThread had two problems.  1. the event came later than the caller is expecting it, and in at least one test the event was disabled before posting, and 2. there's a comment in the code why we can't post events with a JavaThread.  We'd have to transition into native while holding a no safepoint lock (or else deadlock).  The point of making this change was so that the JVMTI table does not need GC code to serially process the table.

kimbarrett replied:
I know of nothing that leads to "presumably during GC" being a requirement. Having all pending events of some type occur before that type of event is disabled seems like a reasonable requirement, but just means that event disabling also requires the table to be "up to date", in the sense that any GC-cleared entries need to be dealt with. That can be handled just like other operations that use the table contents, rather than during the GC.  That is, use post_dead_object_on_vm_thread if there are or might be any pending dead objects, before disabling the event.

fisk replied:
The point of this change was not necessarily to be lazy about updating the tagmap, until someone uses it. The point was to get rid of the last annoying serial GC phase. Doing it all lazily would certainly also achieve that. But it would also lead to situations where no event is ever posted from GC to GC. So you would get the event 20 GCs later, which might come as a surprise. It did come as a surprise to some tests, so it is reasonable to assume it would come as a surprise to users too. And I don't think we want such surprises unless we couldn't deal with them. And we can.

coleenp replied:
Kim's change to post the events from the service thread or before other JVMTI operations removes posting events from  the gc_notification, which was the objection.

dcubed-ojdk commented:
@coleenp - please make sure you hear from someone on the Serviceability team
for this PR...

coleenp commented on src/hotspot/share/prims/jvmtiTagMap.cpp:
```diff
@@ -3298,117 +2975,49 @@ void JvmtiTagMap::follow_references(jint heap_filter,
 }
 
+// Concurrent GC needs to call this in relocation pause, so after the objects are moved
+// and have their new addresses, the table can be rehashed.
```
So the design is that when the oops have new addresses, we set a flag in the table to rehash it.  Not sure why this is wrong and why wouldn't it work for shenandoah? @zhengyu123 ?  When we call WeakHandle.peek()/resolve() after the call, the new/moved oop address should be returned.  Why wouldn't this be the case?

kimbarrett replied:
It used to be that jvmti tagmap processing was all-in-one (in GC weak reference processing, with the weak clearing, dead table entry removal, and rehashing all done in one pass. This change has split that up, with the weak clearing happening in a different place (still as part of the GC's weak reference processing) than the others (which I claim can now be part of the mutator, whether further separated or not).

"Concurrent GC" has nothing to do with whether tagmaps need rehashing.  Any copying collector needs to do so. A non-copying collector (whether concurrent or not) would not. (We don't have any of those in HotSpot today.)  And weak reference clearing (whether concurrent or not) has nothing to do with whether objects have been moved and so the hashing has been invalidated.

There's also a "well known" issue with address-based hashing and generational or similar collectors, where a simple boolean "objects have moved" flag can be problematic, and which tagmaps seem likely to be prone to.  The old do_weak_oops tries to mitigate it by recognizing when the object didn't move.

coleenp commented on src/hotspot/share/prims/jvmtiTagMap.cpp:
```diff
@@ -536,25 +109,64 @@ JvmtiTagMap* JvmtiTagMap::tag_map_for(JvmtiEnv* env) {
+
+  // The table cleaning, posting and rehashing can race for
+  // concurrent GCs. So fix it here once we have a lock or are
+  // at a safepoint.
```
The comment is trying to describe the situation like:
1. mark-end pause (WeakHandle.peek() returns NULL because object A is unmarked)
2. safepoint for heap walk
   2a. Need to post ObjectFree event for object A before the heap walk doesn't find object A.
3. gc_notification - would have posted an ObjectFree event for object A if the heapwalk hadn't intervened

The check_hashmap() function also checks whether the hash table needs to be rehashed before the next operation that uses the hashtable.

Both operations require the table to be locked.

The unlink and post needs to be in a GC pause for reasons that I stated above.  The unlink and post were done in a GC pause so this isn't worse for any GCs.  The lock can be held for concurrent GC while the number of entries are processed and this would be a delay for some applications that have requested a lot of tags, but these applications have asked for this and it's not worse than what we had with GC walking this table in safepoints.

kimbarrett replied:
For the GCs that call the num_dead notification in a pause it is much worse than what we had. As I pointed out elsewhere, it used to be that tagmap processing was all-in-one, as a single serial subtask taken by the first thread that reached it in WeakProcessor processing. Other threads would find that subtask taken and move on to processing oopstores in parallel with the tagmap processing. Now everything except the oopstorage-based clearing of dead entries is a single threaded serial task done by the VMThread, after all the parallel WeakProcessor work is done, because that's where the num-dead callbacks are invoked. WeakProcessor's parallel oopstorage processing doesn't have a way to do the num-dead callbacks by the last thread out of each parallel oopstorage processing. Instead it's left to the end, on the assumption that the callbacks are relatively cheap.  But that could still be much worse than the old code, since the tagmap oopstorage could be late in the order of processing, and so still effectively be a serial subtask after all the parallel subtasks are done or mostly done.

fisk replied:
So both before and now, this task is a single threaded task. The difference is that before that single threaded task could be performed in parallel to other tasks. So if the table is small, you probably won't be able to notice any difference as small table implies not much to do. And if the table is large, you still probably won't be able to notice any difference as a large table implies it will dominate the pause with both the old and new approach. Any difference at all is bounded at 2x processing time, as it was serial both before and after. But now if we have a perfectly medium balanced table, we can at the very worst observe a theoretical 2x worse processing of this JVMTI table. I think that if we truly did care about this difference, and that it is important to keep this code as well performed as possible, then we would not have a serial phase for this at all. The fact that this has been serial suggests to me that it is not a path that is critical, and therefore I don't think optimizing the theoretical max 2x worse processing times for perfectly medium sized JVMTI tag map tables, is worth the hassle. At least I can't see why this would be of any importance.

coleenp commented:
> @coleenp - please make sure you hear from someone on the Serviceability team
> for this PR...

I've asked @sspitsyn to review this.

coleenp commented on src/hotspot/share/gc/shared/weakProcessorPhases.hpp:
```diff
@@ -40,44 +40,14 @@ class WeakProcessorPhases : AllStatic {
 
   typedef void (*Processor)(BoolObjectClosure*, OopClosure*);
```
ok, removed.

coleenp commented on src/hotspot/share/gc/shared/weakProcessorPhases.hpp:
```diff
@@ -40,44 +40,14 @@ class WeakProcessorPhases : AllStatic {
 };
 
-typedef WeakProcessorPhases::Phase WeakProcessorPhase;
+typedef uint WeakProcessorPhase;
```
Ok, so I'm not sure what to do with this:

  enum Phase { 
    // Serial phase.
    JVMTI_ONLY(jvmti)
  
    // Additional implicit phase values follow for oopstorages.
  };

I've removed the only thing in this enum.

coleenp replied:
>Though it might be possible to go even further and eliminate WeakProcessorPhases as a thing separate from OopStorageSet.

This makes sense.  Can we file another RFE for this? I was sort of surprised by how much code was involved so I tried to find a place to stop deleting.

coleenp replied:
I restored the empty enum for Phase and can open a new RFE if there is more code to remove.

sspitsyn commented:
> > @coleenp - please make sure you hear from someone on the Serviceability team
> > for this PR...
> 
> I've asked @sspitsyn to review this.

Yes, I'm reviewing this. Still need another pass.

sspitsyn reviewed (commented):
Hi Coleen,

Wow, there are a lot of simplifications and code removal with this fix!
It looks great in general, just some nits below.
I also wanted to suggest renaming the 'set_needs_processing' to 'set_needs_rehashing'. :)

src/hotspot/share/prims/jvmtiTagMap.hpp:

Nit: Would it better to use a plural form 'post_dead_objects_on_vm_thread'? :
`+  void post_dead_object_on_vm_thread();`

src/hotspot/share/prims/jvmtiTagMap.cpp:

Nit: It'd be nice to add a short comment before the check_hashmap similar to L143 also explaining a difference (does check and post just for one env) with the check_hashmaps_for_heapwalk:
```
122 void JvmtiTagMap::check_hashmap(bool post_events) {
  . . .
143 // This checks for posting and rehashing and is called from the heap walks.
 144 void JvmtiTagMap::check_hashmaps_for_heapwalk() {
```

I'm just curious how this fragment was added. Did you get any failures in testing? :
```
1038   // skip if object is a dormant shared object whose mirror hasn't been loaded
1039   if (obj != NULL &&   obj->klass()->java_mirror() == NULL) {
1040     log_debug(cds, heap)("skipped dormant archived object " INTPTR_FORMAT " (%s)", p2i(obj),
1041                          obj->klass()->external_name());
1042     return;
1043   }
```

Nit: Can we rename this field to something like '_some_dead_found' or '_dead_found'? :
`1186   bool _some_dead;`

Nit: The lines 2997-3007 and 3009-3019 do the same but in different contexts. 
```
2996   if (!is_vm_thread) {
2997     if (num_dead_entries != 0) {
2998       JvmtiEnvIterator it;
2999       for (JvmtiEnv* env = it.first(); env != NULL; env = it.next(env)) {
3000         JvmtiTagMap* tag_map = env->tag_map_acquire();
3001         if (tag_map != NULL) {
3002           // Lock each hashmap from concurrent posting and cleaning
3003           tag_map->unlink_and_post_locked();
3004         }
3005       }
3006       // there's another callback for needs_rehashing
3007     }
3008   } else {
3009     assert(SafepointSynchronize::is_at_safepoint(), "must be");
3010     JvmtiEnvIterator it;
3011     for (JvmtiEnv* env = it.first(); env != NULL; env = it.next(env)) {
3012       JvmtiTagMap* tag_map = env->tag_map_acquire();
3013       if (tag_map != NULL && !tag_map->is_empty()) {
3014         if (num_dead_entries != 0) {
3015           tag_map->hashmap()->unlink_and_post(tag_map->env());
3016         }
3017         // Later GC code will relocate the oops, so defer rehashing until then.
3018         tag_map->_needs_rehashing = true;
3019       }
```
It feels like it can be refactored/simplified, at least, a little bit.
Is it possible to check and just return if (num_dead_entries == 0)?
If not, then, at least, it can be done the same way (except of locking).
Q: Should the _needs_rehashing be set in both contexts?

Also, can we have just one (static?) lock for the whole gc_notification (not per JVMTI env/JvmtiTagMap)? How much do we win by locking per each env/JvmtiTagMap? Note, that in normal case there is just one agent. It is very rare to have multiple agents requesting object tagging and ObjectFree events. It seems, this can be refactored to more simple code with one function doing work in both contexts.

src/hotspot/share/utilities/hashtable.cpp:

Nit: Need space after the '{' :
  `+const int _small_table_sizes[] = {107, 1009, 2017, 4049, 5051, 10103, 20201, 40423 } ;`

src/hotspot/share/prims/jvmtiTagMapTable.cpp:

Nit: Extra space after assert:
  `119   assert (find(index, hash, obj) == NULL, "shouldn't already be present");`

Thanks,
Serguei

sspitsyn commented:
More about possible refactoring of the JvmtiTagMap::gc_notification().
I'm thinking about something like below:

```
void JvmtiTagMap::unlink_and_post_for_all_envs() {
  if (num_dead_entries == 0) {
    return; // nothing to unlink and post
  }
  JvmtiEnvIterator it;
  for (JvmtiEnv* env = it.first(); env != NULL; env = it.next(env)) {
    JvmtiTagMap* tag_map = env->tag_map_acquire();
    if (tag_map != NULL && !tag_map->is_empty()) {
      tag_map->unlink_and_post();
    }
  }
}

void JvmtiTagMap::gc_notification(size_t num_dead_entries) {
  if (Thread::current()->is_VM_thread()) {
    assert(SafepointSynchronize::is_at_safepoint(), "must be");
    unlink_and_post_for_all_envs();
    set_needs_rehashing();
  } else {
    MutexLocker ml(JvmtiTagMap_lock(), Mutex::_no_safepoint_check_flag);
    unlink_and_post_for_all_envs();
    // there's another callback for needs_rehashing
  }
}
```

If we still need a lock per each JvmtiTagMap then it is possible to add this fragment to the unlink_and_post_for_all_envs:
   bool is_vm_thread = Thread::current()->is_VM_thread()
    MutexLocker ml(is_vm_thread ? NULL : lock(), Mutex::_no_safepoint_check_flag);

Then the code above could look like below:

```
void JvmtiTagMap::unlink_and_post_for_all_envs() {
  if (num_dead_entries == 0) {
    return; // nothing to unlink and post
  }
   bool is_vm_thread = Thread::current()->is_VM_thread()
  JvmtiEnvIterator it;
  for (JvmtiEnv* env = it.first(); env != NULL; env = it.next(env)) {
    JvmtiTagMap* tag_map = env->tag_map_acquire();
    if (tag_map != NULL && !tag_map->is_empty()) {
      MutexLocker ml(is_vm_thread ? NULL : lock(), Mutex::_no_safepoint_check_flag);
      tag_map->unlink_and_post();
    }
  }
}

void JvmtiTagMap::gc_notification(size_t num_dead_entries) {
  if (Thread::current()->is_VM_thread()) {
    assert(SafepointSynchronize::is_at_safepoint(), "must be");
    set_needs_rehashing();
  }
  unlink_and_post_for_all_envs();
}
```

coleenp commented:
@sspitsyn  Thank you for reviewing the code.  The gc_notification refactoring is awkward because each table needs a lock and not a global lock. If we find another place in the GCs to call set_needs_rehashing() it might be possible to make gc_notification call two functions with a boolean to decide whether to take the lock.  We're still working on @kimbarrett comments so maybe the notification will change to some new thread and be refactored that way if necessary.  I fixed the code for your other comments.

sspitsyn reviewed (approved):
Thank you for the update, Coleen!
I leave it for you to decide to refactor the gc_notification or not.
Thanks,
Serguei

coleenp commented:
Thanks @sspitsyn .  I'm going to leave the gc_notification code because structurally the two sides of the if statement are different and it's not a long function.  Thank you for reviewing the change.

coleenp commented on src/hotspot/share/prims/jvmtiTagMapTable.hpp:
```diff
@@ -0,0 +1,101 @@
+// Hashtable to record oops used for JvmtiTagMap
+class JvmtiTagMapEntryClosure;
+
+class JvmtiTagMapEntry : public HashtableEntry<WeakHandle, mtServiceability> {
```
So looking at the concurrent hashtable, the entries can be created without saving the hashcode.  I was going to use that at first but didn't want to cut/paste the boilerplate to do so and the jvmti tag map hashtable is always accessed with a lock.  This could be a future RFE if necessary and would also serve to eliminate another ad-hoc hashtable.

coleenp commented:
This change passes tier 7,8 testing.

coleenp commented:
/author kimbarrett
does this work?
I've added two commits from @kimbarrett that defer the ObjectFree posting to the service thread or to a place where it could be removed before posting.
I also remerged and added the call JvmtiTagMap::set_needs_cleaning() to shenandoah which works after merging the latest code from shenandoah.
Testing tiers 1-6 currently.  jvmti/jdi tests pass with G1 and ZGC stress options, and JVMTI tests pass with shenandoah.
Thanks! Coleen

coleenp commented:
/contributor add @kimbarrett 
/contributor add @coleenp

sspitsyn reviewed (approved):
Hi Coleen,
It looks good to me.
Just a couple of nits below.

src/hotspot/share/prims/jvmtiTagMap.cpp:

There is a double-check for _needs_cleaning, so the one at line 136 can be removed:
```
 136   if (_needs_cleaning &&
 137       post_events &&
 138       env()->is_enabled(JVMTI_EVENT_OBJECT_FREE)) {
 139     remove_dead_entries(true /* post_object_free */);
```
```
1158 void JvmtiTagMap::remove_dead_entries(bool post_object_free) {
1159   assert(is_locked(), "precondition");
1160   if (_needs_cleaning) {
1161     log_info(jvmti, table)("TagMap table needs cleaning%s",
1162                            (post_object_free ? " and posting" : ""));
1163     hashmap()->remove_dead_entries(env(), post_object_free);
1164     _needs_cleaning = false;
1165   }
1166 }
```
test/hotspot/jtreg/vmTestbase/nsk/jvmti/AttachOnDemand/attach021/attach021Agent00.cpp:

The change below is not needed as the call to nsk_jvmti_aod_disableEventAndFinish() does exactly the same:
```
-    nsk_jvmti_aod_disableEventAndFinish(agentName, JVMTI_EVENT_OBJECT_FREE, success, jvmti, jni);
+
+    /* Flush any pending ObjectFree events, which may set success to 1 */
+    if (jvmti->SetEventNotificationMode(JVMTI_DISABLE,
+                                        JVMTI_EVENT_OBJECT_FREE,
+                                        NULL) != JVMTI_ERROR_NONE) {
+        success = 0;
+    }
+
+    nsk_aod_agentFinished(jni, agentName, success);
 }
```

coleenp commented on test/hotspot/jtreg/vmTestbase/nsk/jvmti/AttachOnDemand/attach021/attach021Agent00.cpp:
```diff
@@ -69,7 +69,15 @@ Java_nsk_jvmti_AttachOnDemand_attach021_attach021Target_setTagFor(JNIEnv * jni,
 JNIEXPORT void JNICALL
 Java_nsk_jvmti_AttachOnDemand_attach021_attach021Target_shutdownAgent(JNIEnv * jni,
         jclass klass) {
-    nsk_jvmti_aod_disableEventAndFinish(agentName, JVMTI_EVENT_OBJECT_FREE, success, jvmti, jni);
```
@kimbarrett ? I'm not sure why you made this change.  See Serguei's comment.

coleenp commented:
> There is a double-check for _needs_cleaning, so the one at line 136 can be removed:

I want to leave _needs_cleaning at 136 because even though the boolean is checked twice, it doesn't hurt performance and it has a nice symmetry in that function.

I asked Kim about the other change.

Thank you for reviewing, Serguei!

kimbarrett commented:
> test/hotspot/jtreg/vmTestbase/nsk/jvmti/AttachOnDemand/attach021/attach021Agent00.cpp:
> 
> The change below is not needed as the call to nsk_jvmti_aod_disableEventAndFinish() does exactly the same:
> 
> ```
> -    nsk_jvmti_aod_disableEventAndFinish(agentName, JVMTI_EVENT_OBJECT_FREE, success, jvmti, jni);
> +
> +    /* Flush any pending ObjectFree events, which may set success to 1 */
> +    if (jvmti->SetEventNotificationMode(JVMTI_DISABLE,
> +                                        JVMTI_EVENT_OBJECT_FREE,
> +                                        NULL) != JVMTI_ERROR_NONE) {
> +        success = 0;
> +    }
> +
> +    nsk_aod_agentFinished(jni, agentName, success);
>  }
> ```

This change really is needed.

The success variable in the test is a global, initially 0, set to 1 by the
ObjectFree handler.

In the old code, if the ObjectFree event hasn't been posted yet, we pass the
initial 0 value of success to nsk_jvmti_aod_disabledEventAndFinish, where
it's a local variable (so unaffected by any changes to the variable in the
test), so stays 0 through to the call to nsk_aod_agentFinished.  So the test
fails. 

The split in the change causes the updated post-ObjectFree event success
value of 1 to be passed to agentFinished.  So the test passes.

That required some head scratching to find at the time.  That's the point of
the comment about flushing pending events.  Maybe the comment should be
improved.

sspitsyn commented:
@kimbarrett
Okay, thank you for explanation.
I agree, it'd make sense to improve the comment a little bit.
Thanks,
Serguei

coleenp commented:
So should nsk_jvmti_aod_disableEventAndFinish pass the address of success instead?  Why didn't it fail before this change?

coleenp commented:
/* Flush any pending ObjectFree events, which will set global success variable to 1
       for any pending ObjectFree events. */
How about this?   The word 'global' helps me.

coleenp commented:
With remerging into shenandoah, all the jdi tests pass with shenandoah also.

coleenp commented:
Thank you to all the reviewers.
/integrate
